### PR TITLE
Add support for Normalized IODs and Merging Services Specification

### PR DESCRIPTION
**Pull request by dwikler**
- Improve DOMTableSpecParser to handle module tables with different number of columns
- Update iodattributes CLI app to support both C-IODs and N-IODs
- Add upsioddimseattributes CLI app to combine UPS N-IOD and UPS DIMSE specifications
- Update logging and debug output for better traceability
- Update unit tests for SpecModel and SpecMerger changes
- Fix regression on exclude_titles after change in empty columns handling
- Fix issue in strip_module_level
- Add patching of UPS attributes specification table

## Summary by Sourcery

Support Normalized IODs and UPS DIMSE service merging with enhanced table parsing and merge logic

New Features:
- Add support for parsing and merging Normalized IODs (N-IODs) in addition to Composite IODs
- Introduce a new upsioddimseattributes CLI to combine UPS I-IOD and UPS DIMSE specifications
- Add merge_path_with_default in SpecMerger to merge models and assign default values for missing attributes
- Add ignore_module_level option to SpecMerger path-based merging to strip module nesting for matching

Bug Fixes:
- Fix regression in exclude_titles after empty-column handling change
- Correct strip_module_level logic in path matching to properly handle module levels
- Fix attribute detection in SpecModel._has_only_key_0_attr to treat missing attributes as None

Enhancements:
- Extend DOMTableSpecParser to handle extra or missing columns, skip empty rows/cells, respect skip_columns, and realign headers
- Refactor SpecMerger.merge_many and internal merge logic to require explicit match_by/attribute_names, validate inputs, and add missing nodes from secondary models
- Improve SpecFactory to accept parser_kwargs (e.g. skip_columns) and propagate them to the parser
- Enhance logging and debug output across parsing, merging, and default-value application for better traceability
- Update SpecPrinter to skip rows with all empty values

Tests:
- Add tests for DOMTableSpecParser (extra/fewer columns, empty rows/cells, colspan, mixed colspan, skip_columns)
- Add tests for SpecMerger (merge_path_with_default defaults, ignore_module_level behavior, add_missing_nodes, merge_many validations)
- Update tests for SpecFactory, IODSpecBuilder, SpecModel, SpecPrinter, CSV parser, and CLI behaviors

### CODE REVIEW

**Comment by dwikler** [comment]
> ```
> Please address the comments from this code review:
> ## Overall Comments
> - The new merge_path_with_default method duplicates much of merge_path’s setup—consider extracting the shared merge-initialization logic into a helper to keep defaults and docstrings in sync.
> ```

The shared setup between `merge_path` and `merge_path_with_default` is already encapsulated by having `merge_path_with_default` call `merge_path` directly. Introducing an additional helper would not reduce duplication or improve maintainability in this case.

**Comment by dwikler** [comment]
> * The branching inside DOMTableSpecParser._extract_row_data for skip_columns makes the method quite complex—refactor the column‐to‐cell alignment into a smaller utility or separate methods for better readability.

I've refactored the column-to-cell alignment logic in _extract_row_data into two private methods: _align_row_with_skipped_columns and _align_row_default.

File: src/dcmspec/cli/upsioddimseattributes.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    ups_model.select_role(args.role)
+
+    # --- Merge by path with DICOM service default type logic ---
+
+    # Use UPS_DIMSE_MAPPING to get the attributes to merge for the selected DIMSE
+    dimse_info = UPS_DIMSE_MAPPING.get(args.dimse, {})
+    dimse_attributes = dimse_info.get("attributes", [])
+    dimse_req_attributes = dimse_info.get("req_attributes", [])
+    # Add "comment" to the end of the list
+    dimse_attributes.append("comment")
```
[the remark this answers is not in the document]
Comment (dwikler) [reply src/dcmspec/cli/upsioddimseattributes.py:245]: In this context, the attributes list is sourced directly from UPS_DIMSE_MAPPING, which I control and know does not already include "comment". Therefore, appending "comment" will not cause duplicates, and an explicit check is not necessary here.

File: tests/test_spec_model.py [hunk 2]
```diff
@@ -166,6 +168,87 @@ def test_merge_matching_path_by_attribute(merge_by_path_test_models):
...
+
+    # Print merged tree and attributes for debugging (captured by capsys)
+    def get_real_path(node):
+        # Use elem_tag if present, otherwise name (for all levels)
+        path = []
+        for n in node.path:
+            if hasattr(n, "elem_tag") and getattr(n, "elem_tag", None) is not None:
+                path.append(getattr(n, "elem_tag"))
+            else:
+                path.append(getattr(n, "name", None))
```
[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_spec_model.py:193]: Removed the debug code block including the loop.

[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_spec_model.py:193]: Removed the debug code block including the conditional.

File: tests/test_spec_model.py [hunk 3]
```diff
@@ -166,6 +168,87 @@ def test_merge_matching_path_by_attribute(merge_by_path_test_models):
...
+        return tuple(path)
+
+    print("Current model real paths:")
+    for node in current.content.descendants:
+        print(get_real_path(node))
```
[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_spec_model.py:198]: Removed the debug code block including the loop.

File: tests/test_spec_model.py [hunk 4]
```diff
@@ -166,6 +168,87 @@ def test_merge_matching_path_by_attribute(merge_by_path_test_models):
...
+    
+    print("Other model real paths:")
+    for node in other.content.descendants:
+        print(get_real_path(node))
```
[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_spec_model.py:202]: Removed the debug code block including the loop.

File: tests/test_spec_model.py [hunk 5]
```diff
@@ -166,6 +168,87 @@ def test_merge_matching_path_by_attribute(merge_by_path_test_models):
...
+
+    print("Current model paths:")
+    for node in current.content.descendants:
+        print(SpecModel._get_path_by_attr(node, "elem_tag"))
+
+    print("Other model paths:")
+    for node in other.content.descendants:
+        print(SpecModel._get_path_by_attr(node, "elem_tag"))
```
[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_spec_model.py:210]: Removed the debug code block including the loop.

File: src/dcmspec/dom_table_spec_parser.py [hunk 6]
```diff
@@ -336,15 +393,28 @@ def _create_node(
...
-        header = [header.get_text(strip=True) for i, header in enumerate(cells) if i in column_to_attr]
+        num_columns = len(cells)
+        # If the mapping has non-consecutive keys and the table has fewer columns, realign
+        if max(column_to_attr.keys()) >= num_columns:
+            # Map consecutive indices to the same attribute names, skipping as needed
+            sorted_attrs = [column_to_attr[k] for k in sorted(column_to_attr.keys())]
+            realigned_col_to_attr = {i: attr for i, attr in enumerate(sorted_attrs)}
+            column_to_attr = realigned_col_to_attr
+
+        header = []
```
[the remark this answers is not in the document]
Comment (dwikler) [reply src/dcmspec/dom_table_spec_parser.py:414]: Fixed accordingly.

File: tests/test_dom_table_spec_parser.py [hunk 7]
```diff
@@ -103,10 +105,67 @@ def test_parse_returns_metadata_and_content(docbook_sample_dom_1):  # noqa: F811
...
+    children = list(node.children)
+    assert len(children) == 2
+    assert hasattr(children[0], "elem_name")
+    assert hasattr(children[0], "elem_tag")
+    assert not hasattr(children[0], "elem_type")
+    assert not hasattr(children[0], "elem_desc")
+    assert hasattr(children[1], "elem_name")
+    assert hasattr(children[1], "elem_tag")
+    assert not hasattr(children[1], "elem_type")
+    assert not hasattr(children[1], "elem_desc")
```
[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_dom_table_spec_parser.py:147]: For small, simple assertion blocks that only appear twice, the trade-off of indirection and reduced clarity does not seem justified. I will skip this comment.

File: tests/test_spec_model.py [hunk 8]
```diff
@@ -21,29 +21,31 @@ def test_init_sets_metadata_and_content_and_logger(simple_spec_model):
...
+    setattr(title_node, "elem_tag", None)
+    # Non-title node: both attributes set
     non_title_node = Node("my_element", parent=content)
     setattr(non_title_node, "elem_name", "MyElement")
     setattr(non_title_node, "elem_tag", "(0101,0010)")
-    # This node should be considered an include (name contains "include_table")
+    # Include node: name contains "include_table"
     include_node = Node("include_table_macro", parent=content)
     setattr(include_node, "elem_name", "Include Table Macro")
+    setattr(include_node, "elem_tag", None)
```
[the remark this answers is not in the document]
Comment (dwikler) [reply tests/test_spec_model.py:40]: For small, simple assertion blocks that only appear twice, the trade-off of indirection and reduced clarity does not seem justified. I will skip this comment.
